src/monodroid/jni/monodroid-glue.cc

```diff
@@ -1219,145 +1202,176 @@ MonodroidRuntime::set_environment_variable_for_directory (const char *name, jstr
 }
 
 #if DEBUG
 void
 MonodroidRuntime::set_debug_env_vars (void)
 {
-	char *value;
-
-	if (androidSystem.monodroid_get_system_property (Debug::DEBUG_MONO_ENV_PROPERTY, &value) == 0)
+	dynamic_local_string<PROPERTY_VALUE_BUFFER_LEN> value;
+	if (androidSystem.monodroid_get_system_property (Debug::DEBUG_MONO_ENV_PROPERTY, value) == 0)
 		return;
 
-	char **args = utils.monodroid_strsplit (value, "|", 0);
-	delete[] value;
+	auto log_envvar = [](const char *name, const char *v) {
+		log_info (LOG_DEFAULT, "Env variable '%s' set to '%s'.", name, v);
+	};
+
+	string_segment arg_token;
+	while (value.next_token ('|', arg_token)) {
+		static_local_string<SMALL_STRING_PARSE_BUFFER_LEN> arg (arg_token.length ());
+		arg.assign (arg_token.start (), arg_token.length ());
 
-	for (char **ptr = args; ptr && *ptr; ptr++) {
-		char *arg = *ptr;
-		char *v = strchr (arg, '=');
-		if (v) {
-			*v = '\0';
-			++v;
-		} else {
+		ssize_t idx = arg.index_of ('=');
+		size_t index = static_cast<size_t>(idx);
+		if (idx < 0 || index == arg.length () - 1) {
+			// ’name’ or ’name=’
 			constexpr char one[] = "1";
-			v = const_cast<char*> (one);
+			if (index == arg.length () - 1) {
+				arg[index] = '\0';
+			}
+			setenv (arg.get (), one, 1);
+			log_envvar (arg.get (), one);
+		} else if (index == 0) {
+			// ’=value’
+			log_warn (LOG_DEFAULT, "Attempt to set environment variable without specifying name: '%s'", arg.get ());
+		} else {
+			// ’name=value’
+			arg[index] = '\0';
+			const char *v = arg.get () + idx + 1;
+			setenv (arg.get (), v, 1);
+			log_envvar (arg.get (), v);
 		}
-		setenv (arg, v, 1);
-		log_info (LOG_DEFAULT, "Env variable '%s' set to '%s'.", arg, v);
 	}
-	utils.monodroid_strfreev (args);
 }
 #endif /* DEBUG */
 
 inline void
 MonodroidRuntime::set_trace_options (void)
 {
-	char *value;
-
-	if (androidSystem.monodroid_get_system_property (Debug::DEBUG_MONO_TRACE_PROPERTY, &value) == 0)
+	dynamic_local_string<PROPERTY_VALUE_BUFFER_LEN> value;
+	if (androidSystem.monodroid_get_system_property (Debug::DEBUG_MONO_TRACE_PROPERTY, value) == 0)
 		return;
 
-	mono_jit_set_trace_options (value);
-	delete[] value;
+	mono_jit_set_trace_options (value.get ());
 }
 
 inline void
 MonodroidRuntime::set_profile_options ()
 {
-	constexpr const char mlpd_ext[] = "mlpd";
-	constexpr const char aot_ext[] = "aotprofile";
+	log_warn (LOG_DEFAULT, __PRETTY_FUNCTION__);
+	// We want to avoid dynamic allocation, thus let’s create a buffer that can take both the property value and a
+	// path without allocation
+	dynamic_local_string<SENSIBLE_PATH_MAX + PROPERTY_VALUE_BUFFER_LEN> value;
+	{
+		dynamic_local_string<PROPERTY_VALUE_BUFFER_LEN> prop_value;
+		if (androidSystem.monodroid_get_system_property (Debug::DEBUG_MONO_PROFILE_PROPERTY, prop_value) == 0)
+			return;
 
-	constexpr const char output_arg[] = "output=";
-	constexpr const size_t output_arg_len = sizeof(output_arg) - 1;
+		value.assign (prop_value.get (), prop_value.length ());
```

Comment:
Seeing this statement makes me wonder if there should instead be an `operator=()` or a "move" operator for `value = std::move(prop_value)`;

Reply:
Not sure if move semantics makes sense since the local buffers are by definition ephemeral so there's no need to add an overhead of implementing proper move semantics (the standard requires that the object moved from is left in some valid state but that the data is gone - that means extra assignments and checks while the object is going away with the scope end anyway). The `=` operator can't take the length parameter and it would become merely a wrapper for `assign(const char*)` - syntactic sugar, not much more. I can add it if you want, but it wouldn't be used here anyway since the `strlen` it would incur is simply not necessary given that `prop_value.length()` does the job without scanning the string for the `NUL` terminator.